feat: Is this this ok?

phoevos opened the pull request:
Did I fix the branch or did I just make it worse?

xiatianrui1110 committed:
rebase evaluation commits

GrayNekoBean committed:
fix CLI import

phoevos committed:
eval: Update Python requirements and project config

* Add requirements.in and generate requirements.txt for evaluation
  scripts
* Update requirements-dev.txt to include evaluation requirements
* Fix backend requirements to avoid conflicts with evaluation
* Update pyproject.toml to include evaluation scripts in linting

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

binxuan39 committed:
Add a README.md and fix naming bug

xiatianrui1110 committed:
deal with comments

xiatianrui1110 committed:
deal with comments and modify the readme file

xiatianrui1110 committed:
fix readme format

xiatianrui1110 committed:
fix readme format

binxuan39 committed:
Update Red Teaming

- Allow user to set the output directory
- Allow to run the redteaming on all the scorer in the directory in one time
- Upadte README.md regarding the running of red teaming

binxuan39 committed:
Update README.md with the instruction of generating answers from questions

phoevos committed:
fix: Very Important Changes

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

binxuan39 committed:
Update red_teaming.py, adding validation of input yaml file

phoevos committed:
fixup! fix: Very Important Changes

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: Very Important Changes

Fix `send_question_to_target` and relevant tests.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Ignore PyRIT on unsupported Python versions

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: Ignore PyRIT on unsupported Python versions

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Pin azure-ai-generative[evaluate]

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
Merge remote-tracking branch 'origin/main' into evaluation-rb

phoevos committed:
fixup! fix: Ignore PyRIT on unsupported Python versions

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Add tests for evaluate.py

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
Merge branch 'main' into evaluation-rb

xiatianrui1110 committed:
add target class for red teaming and update readme

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+EVALUATION_DIR = Path(__file__).parent
+
+logger = logging.getLogger(__name__)
```
Everywhere else we use the package name (i.e. "evaluation") instead of the module one which you're using here. Let's be consistent. We usually go for the module name btw.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+load_dotenv()
+
+EVALUATION_DIR = Path(__file__).parent
```
Why are we redefining this? Why are we not importing it? Why are we not receiving it as a param where needed?

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+from evaluation.utils import load_config
+
+load_dotenv()
```
Why are we loading the environment again? This should only be done once in our application entrypoint (e.g. `cli.py`). What's more important is that while the first time you call `load_dotenv` you choose to override the system envvars, in this call you aren't.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+    PromptRequestResponse,
+    construct_response_from_request,
+)
+from pyrit.prompt_target import PromptChatTarget
```
Please look at my previous changes and understand why these break the tests for Python 3.9 and 3.12

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
```
We don't have this anywhere else.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+class AppChatTarget(PromptChatTarget):
+
+    BACKEND_URI: str = os.environ.get("BACKEND_URI", "").rstrip("/") + "/ask"
```
This is a class attribute for a class called `AppChatTarget`. `BACKEND_URI` is a name that makes sense as an envvar and only in the context of our deployment. This class doesn't have anything to do with that, all it cares about is talking to an application and getting a response. If we do need this attribute here, I would expect it to be name sth like "endpoint" for instance.

phoevos replied:
Also, what are you doing there? If the envvar doesn't exist does it make sense for you to use an empty string as the default? What are you gonna do with it? And why would you go on to concatenate it with `/ask`?

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+        self.endpoint_uri: str = default_values.get_required_value(
+            env_var_name=self.BACKEND_URI, passed_value=endpoint_uri
+        )
```
What are you doing here? Please check how `default_values.get_required_value` works. It gets an envvar name (e.g. "BACKEND_URI") and a passed value. Its implementation is terribly simple:
```python
if passed_value:
        return passed_value

value = os.environ.get(env_var_name)
if value:
        return value
```
In your case, if an endpoint URI is provided, it will be used. If not, the function will try to fetch an environment variable called sth like `https://app-backend-xm5gmlddfm3cc.azurewebsites.net/ask`. Of course, such a thing doesn't exist (the name of the envvar is `BACKEND_URI`).

But all of these are beside the point. This is how one handles defaulting in python:
```python
self.endpoint_uri: str = endpoint_uri or self.BACKEND_URI
```

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        if not resp_text:
+            raise ValueError("The chat returned an empty response.")
+
+        logger.info(f'Received the following response from the prompt target "{resp_text}"')
```
We default to using double quotes for strings in Python (see everywhere else)!

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        app_config = load_config(config)
+        squashed_messages = self.chat_message_normalizer.normalize(messages)
+        messages_dict = [message.model_dump() for message in squashed_messages]
+        target_parameters = app_config.get("target_parameters", {})
```
All we need from the config are the target parameters. Why not pass them directly when initialising the `AppChatTarget` object? The targets they implement for instance receive stuff like `temperature`, `max_tokens`, `top_k` on init. We can do sth similar, by providing a single dict param, `target_parameters`, so that we can cater for any arbitrary model.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+
+        return headers
+
+    def _validate_request(self, *, prompt_request: PromptRequestResponse) -> None:
```
Why did you get rid of the docstrings? I've added them everywhere else in the `evaluation` script.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        return data
+
+    def _get_headers(self) -> dict:
+
```
Why did you get rid of the docstrings? I've added them everywhere else in the `evaluation` script.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        config: Path = EVALUATION_DIR / "config.json"
+        app_config = load_config(config)
+        squashed_messages = self.chat_message_normalizer.normalize(messages)
+        messages_dict = [message.model_dump() for message in squashed_messages]
```
@nextedoff is this ok here? I remember the whole discussion about dataclasses and dicts when it comes to dumping the model.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        self,
+        messages: list[ChatMessage],
+    ) -> str:
+
```
Why did you get rid of the docstrings? I've added them everywhere else in the `evaluation` script.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        self.chat_message_normalizer = chat_message_normalizer
+
+    async def send_prompt_async(self, *, prompt_request: PromptRequestResponse) -> PromptRequestResponse:
+
```
Why did you get rid of the docstrings? I've added them everywhere else in the `evaluation` script.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        )
+
+        if not resp_text:
+            raise ValueError("The chat returned an empty response.")
```
The original code wraps this with error handling for dealing with bad requests and rate limiting. Why don't we need that?

phoevos replied:
Fixed.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+            endpoint_uri=self.endpoint_uri, method="POST", request_body=payload, headers=headers
+        )
+
+        return response.json()["message"]["content"]
```
Do we need to check whether these fields exist? Are they always the same?

phoevos commented on evaluation/evaluate.py:
```diff
@@ -140,7 +141,7 @@ def run_evaluation(
     with open(results_dir / "evaluate_parameters.json", "w", encoding="utf-8") as parameters_file:
         parameters = {
-            "evaluation_gpt_model": openai_config.get("model", "unknown_model"),
+            "evaluation_gpt_model": openai_config.model,
```
Why?

xiatianrui1110 replied:
> Why?

Because it showed an error of "AttributeError: 'AzureOpenAIModelConfiguration' object has no attribute 'get'"

phoevos replied:
Makes sense, doesn't it, because you've changed the `openai_config` from a `dict` (which supports "get") to a `ModelConfiguration` object. I've asked elsewhere why that decision was made.

phoevos commented on evaluation/red_teaming.py:
```diff
@@ -62,7 +62,7 @@ async def run_red_teaming(
         ) as red_teaming_orchestrator:
             score = await red_teaming_orchestrator.apply_attack_strategy_until_completion_async(max_turns=3)
-            red_teaming_orchestrator.print_conversation()
+            # red_teaming_orchestrator.print_conversation()
```
Let's not leave commented out code behind.

phoevos commented on evaluation/service_setup.py:
```diff
@@ -170,6 +172,13 @@ def get_openai_target() -> PromptChatTarget:
 
+def get_app_target() -> PromptChatTarget:
+    """Get specified OpenAI chat target."""
+    endpoint = os.environ.get("BACKEND_URI", "").rstrip("/") + "/ask"
```
Same comment as in the class attribute case. Isn't this a required param? Why are you using an empty string as the default? What's the point of continuing if a URI is not provided?

phoevos commented on evaluation/service_setup.py:
```diff
@@ -170,6 +172,13 @@ def get_openai_target() -> PromptChatTarget:
 
 
+def get_app_target() -> PromptChatTarget:
+    """Get specified OpenAI chat target."""
```
Not OpenAI anymore.

phoevos commented on evaluation/evaluate.py:
```diff
@@ -31,6 +31,7 @@ def send_question_to_target(question: str, url: str, parameters: dict = {}, rais
         "context": parameters,
     }
     try:
+        print(url)
```
Looks like a print used for debugging, let's remove it.

phoevos commented on evaluation/cli.py:
```diff
@@ -97,12 +97,15 @@ def red_teaming(
-    target = (
-        service_setup.get_openai_target() if prompt_target == "openai" else service_setup.get_azure_ml_chat_target()
-    )
+    if prompt_target == "application":
+        target = service_setup.get_app_target()
+    elif prompt_target == "azureopenai":
+        target = service_setup.get_openai_target()
+    elif prompt_target == "azureml":
+        target = service_setup.get_azure_ml_chat_target()
```
What if it's none of these? Nowhere do we mention that this is an enumeration or that it can only take specific values. That should be exposed through the CLI, definitely as a help message. And we should be handling the case where the provided value doesn't match any of these. Not only is an error not being returned here, but you proceed with an undefined target to run the red teaming.

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        headers = self._get_headers()
+        payload = self._construct_http_body(messages)
+
+        response = await net_utility.make_request_and_raise_if_error_async(
+            endpoint_uri=self.endpoint_uri, method="POST", request_body=payload, headers=headers
+        )
```
Not a huge fan of using a util from an external library for something as simple as that, but ok :rofl:

phoevos commented on evaluation/app_chat_target.py:
```diff
@@ -0,0 +1,111 @@
+        self,
+        *,
+        endpoint_uri: str = None,
+        chat_message_normalizer: ChatMessageNormalizer = ChatMessageNop(),
```
Note that we're not using this. We're not passing a normaliser on initialisation currently and the default one simply returns the messages. I'm not saying that we should remove it, just raising it FYI.

phoevos reviewed (commented):
PTAL, some changes are needed! Have we tested that with Hugging Face?

GrayNekoBean committed:
added request failing check for evaluation

phoevos commented on evaluation/evaluate.py:
```diff
@@ -30,12 +31,17 @@ def send_question_to_target(question: str, url: str, parameters: dict = {}, rais
         latency = r.elapsed.total_seconds()
-
+                
+        if r.status_code.__str__()[0] != '2':
+            raise ConnectionError(
+                f"Request to target {url} failed with status code {r.status_code}: {r.reason} \n"
+            )
```
What is that? Note btw that we never call under methods directly, instead you would do `str(r.status_code)`. But what are you checking for exactly?

GrayNekoBean replied:
> we never call under methods directly, instead you would do str(r.status_code). 

yes you right, I didn't notice that, I'll change that to `str(r.status_code)`. 

> But what are you checking for exactly?

This code will check if the response to the request is returned with a success status code before parsing JSON. Because otherwise, even when the request to API returned with an error, this function will still process forward and attempt to parse the response body as JSON anyway, and got a "Value Error" that hints the error is occurred when parsing JSON, which is misleading.

And I also set the default value of the parameter `raise_error` to true. Because I think in most cases we want to explicitly throw the error instead of hiding it in result file values when meeting such kind of severe exception.

phoevos replied:
But what does this check mean exactly?

```python
str(r.status_code)[0] != "2"
```

phoevos replied:
Oh wow, I think I see it, you're checking whether the status code is `2xx`.

phoevos replied:
Mate, this is absolutely horrendous :rofl:

GrayNekoBean replied:
Yes that's right

GrayNekoBean committed:
request failing check fixup

- changed a way to check if the returned status code is 2xx

GrayNekoBean committed:
added request failed test and fixed multiple outdated evaluation tests

- added a test case that tests when a request returns failed in method `test_send_question_to_target`.
- parameter `openai_config` will not be passed as a dict any more, instead a AzureOpenAIModelConfiguration object will be passed to methods that are under testing.
- due to raise_error is now set to True by default, some unexpected Exception in tests that was originally caught by try statement in `send_question_to_target` is now addressed.

xiatianrui1110 committed:
deal with comments

xiatianrui1110 committed:
resolve conflict

GrayNekoBean committed:
fix request check

GrayNekoBean committed:
fix documents format

- format test_evaluate.py with ruff
- fix markdown lint for evaluation/README.md

GrayNekoBean committed:
evaluate test fix

- added `ok` property for mock response

xiatianrui1110 committed:
Merge branch 'evaluation-rb' of github.com:ucl-contoso-chat/ucl-openai-search into evaluation-rb
resolve conflicts

phoevos commented on evaluation/evaluate.py:
```diff
@@ -0,0 +1,273 @@
+        r.encoding = "utf-8"
+        latency = r.elapsed.total_seconds()
+
+        if not r.ok:
+            raise ConnectionError(f"Request to target {url} failed with status code {r.status_code}: {r.reason} \n")
```
@GrayNekoBean what you're trying to do there already exists in the library. `response.raise_for_status()` will raise the error that caused the request to be unsuccessful. In fact, `ok` is [using it under the hood](https://github.com/psf/requests/blob/0e322af87745eff34caffe4df68456ebc20d9068/src/requests/models.py#L755-L767). What's more though, this is not a `ConnectionError`, it is specifically an `HTTPError`. It's already there, why would you raise a different error than the one the library returned?

GrayNekoBean replied:
Thx, I still have a lot to learn.

GrayNekoBean committed:
update send_question_to_target

- fixed http request fail checking

phoevos commented on tests/test_evaluate.py:
```diff
@@ -171,11 +171,37 @@ def test_run_evaluation():
         self.elapsed = timedelta(seconds=1)
+        self.url = url
+
+    def raise_for_status(self):
+        """Raises :class:`HTTPError`, if one occurred."""
+
+        http_error_msg = ""
+        if isinstance(self.reason, bytes):
+            # We attempt to decode utf-8 first because some servers
+            # choose to localize their reason strings. If the string
+            # isn't utf-8, we fall back to iso-8859-1 for all other
+            # encodings. (See PR #3538)
+            try:
+                reason = self.reason.decode("utf-8")
+            except UnicodeDecodeError:
+                reason = self.reason.decode("iso-8859-1")
+        else:
+            reason = self.reason
+
+        if 400 <= self.status_code < 500:
+            http_error_msg = f"{self.status_code} Client Error: {reason} for url: {self.url}"
+
+        elif 500 <= self.status_code < 600:
+            http_error_msg = f"{self.status_code} Server Error: {reason} for url: {self.url}"
+
+        if http_error_msg:
+            raise requests.HTTPError(http_error_msg, response=self)
```
@GrayNekoBean why do we copy-paste the [actual implementation of the method](https://github.com/psf/requests/blob/0e322af87745eff34caffe4df68456ebc20d9068/src/requests/models.py#L997-L1024) in a mocked response? If we wanted the actual thing to run, we wouldn't really mock it, would we?

On a side note, when copy-pasting stuff like that, I expect to find an explicit reference in the commit message and possibly in a code comment, pointing to the source.

GrayNekoBean replied:
When I committed it I realized you might have some comment against these lines, but I couldn't thought of any better ideas. My thought is to mock how it actually behaves as a Response object even it's a mocked one. I'll add a comment here that states the original source of this snippet in next commit.

binxuan39 committed:
Update red teaming, allowing customized conversation objective

binxuan39 committed:
fix: make conversation_objective field of red teaming optional

xiatianrui1110 committed:
skip python test

xiatianrui1110 committed:
Merge branch 'evaluation-rb' of github.com:ucl-contoso-chat/ucl-openai-search into evaluation-rb

binxuan39 committed:
Allow red teaming to use pyrit built-in true-false scorer, Update README.md

binxuan39 committed:
Update README.md formatting

GrayNekoBean committed:
fix evaluate test mock response

phoevos committed:
fix: Rename skip Python decorator

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: make conversation_objective field of red teaming optional

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix format but correctly this time

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Only load dotenv once

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Default to using double quotes for strings

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make endpoint_uri compulsory for chat target

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update docstrings in app_chat_target.py

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add retry decorator to app_chat_target.py

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Refactor list of mocked modules in conftest.py

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Check if message content exists

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Pass config dict to get_app_target

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Wrap target request with error handling

This is not ideal because we're using a PyRIT util to send out the
request which means that we subsequently have to catch PyRIT exceptions
as well as rely on httpx for the actual request.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update scorer definitions

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Check if class_identifier exists for scorer

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Add util for saving JSONL files

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Add helper for generating evaluation answers

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Expose generate_answers function to the CLI

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Add scripts for creating evaluation .env

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Remove stale JMESPath refs from config.json

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update list of expected environment variables

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Pretty dump scores.json

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Standardise handling of targeturl

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Convert docstring to imperative

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Drop evaluate_parameters.json

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update README.md

* Update README.md to reflect the changes in the project
* Rephrase some parts
* Standardise the formatting of the provided CLI commands
* Use 'create_eval_dotenv' scripts to set up the environment
* Remove '.env.sample' and 'load_env' scripts which are no longer needed

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: Update README.md

* Fix stupid markdown mistakes

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update .gitignore for evaluation/output

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>
